subsys/net/l2/ethernet/ethernet.c

```diff
@@ -431,7 +440,7 @@ static enum net_verdict ethernet_send(struct net_if *iface,
 		}
 
 		/* Trying to send ARP message so no need to setup it twice */
-		if (ntohs(NET_ETH_HDR(pkt)->type) != NET_ETH_PTYPE_ARP) {
```

Comment:
Thanks for the updates. Unfortunately, I still don't understand what's the problem, even after looking thru your changes and old code for ~10min. Is it the case that pkt can arrive here without Ethernet headers, so trying to access it (`NET_ETH_HDR(pkt)`) means accessing invalid data? It's hard to believe that, because line before this code https://github.com/zephyrproject-rtos/zephyr/pull/9350/files#diff-b7cca9bcea20a1b9bfe87f98f5e65feeL427 accesses NET_ETH_HDR(pkt) anyway.

Reply:
IPv4 prepares ARP message and then sends it. The ethernet ARP routine should not try to prepare a new ARP message so that is the reason for the check in ethernet_send(). My earlier check was obviously wrong but for yet some unknown reason it worked ok for ping, I will investigate this ping issue separately.

Reply:
> IPv4 autoconf prepares ARP message and then sends it. The ethernet ARP routine should not try to prepare a new ARP message for this existing ARP msg

Compare this to what's written in the commit message:

> The ethernet sending routine sent a corrupted ARP packet instead
of the actual IPv4 packet.

Anyway, let me approve this if this fixes the matters.